src/WebJobs.Extensions.DocumentDB/Bindings/DocumentDBAttributeBindingProvider.cs

```diff
@@ -0,0 +1,131 @@
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            ParameterInfo parameter = context.Parameter;
+            DocumentDBAttribute attribute = parameter.GetCustomAttribute<DocumentDBAttribute>(inherit: false);
+            if (attribute == null)
+            {
+                return null;
+            }
+
+            if (_docDBConfig.EndpointUri == null)
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.CurrentCulture,
+                    "The DocumentDB Endpoint Uri must be set either via a '{0}' app setting or directly in code via DocumentDBConfiguration.EndpointUri.",
+                    DocumentDBConfiguration.AzureWebJobsDocumentDBUriName));
+            }
+
+            if (string.IsNullOrEmpty(_docDBConfig.AuthKey))
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.CurrentCulture,
+                    "The DocumentDB Authentication Key must be set either via a '{0}' app setting or directly in code via DocumentDBConfiguration.AuthKey.",
+                    DocumentDBConfiguration.AzureWebJobsDocumentDBKeyName));
+            }
+
+            DocumentDBContext documentDBContext = CreateContext(_docDBConfig, attribute, _jobHostConfig.NameResolver);
+
+            await CreateIfNotExists(attribute, documentDBContext);
+
+            Type coreType = DocumentDBUtility.GetCoreType(parameter.ParameterType);
+
+            IBinding binding = null;
+            try
+            {
+                binding = BindingFactory.BindGenericCollector(parameter, typeof(DocumentDBAsyncCollector<>), coreType,
+                _jobHostConfig.GetService<IConverterManager>(), (s) => documentDBContext);
+            }
+            catch (Exception ex)
+            {
+                // BindingFactory uses a couple levels of reflection so we get TargetInvocationExceptions here.
+                // This pulls out the root exception and rethrows it.
+                while (ex.InnerException != null)
+                {
+                    ex = ex.InnerException;
+                }
+
+                ExceptionDispatchInfo.Capture(ex).Throw();
+            }
+
+            return binding;
+        }
+
+        internal static DocumentDBContext CreateContext(DocumentDBConfiguration config, DocumentDBAttribute attribute, INameResolver resolver)
+        {
+            return new DocumentDBContext
+            {
+                Service = new DocumentDBService(config.EndpointUri, config.AuthKey),
+                ResolvedDatabaseName = Resolve(attribute.DatabaseName, resolver),
+                ResolvedCollectionName = Resolve(attribute.CollectionName, resolver)
+            };
+        }
+
+        internal static async Task CreateIfNotExists(DocumentDBAttribute attribute, DocumentDBContext context)
```

Comment:
This doesnt feel quite right, you're breaking the naming convention by having an async method that does not have an async prefix. Not sure exactly what you should change it to.. CreateIfNotExistsAndSettingIsEnabledAsync is a bit too wordy...